text/2021-12-09-logging.md

```diff
@@ -0,0 +1,302 @@
+# Logging
+
+## Summary
+
+Refine the guide for logging for developing Chaos Mesh.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+Logging in each component in Chaos Mesh is bad. Related issue:
+https://github.com/chaos-mesh/chaos-mesh/issues/2149. Because of the mess on
+logging, the debugging and profiling of the components are difficult.
+
+This proposal aims to improve the logging observability of the components and
+the development experience on printing logs.
+
+After passed this proposal, I would write a developing guide for logging in the
+repo of Chaos Mesh.
+
+## Requirements
+
+- structured logging
+- leveled logging
+- runtime level configure
+- could logging in everywhere (every line of code)
+
+## Detailed design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail that:
+
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+- How the feature is used. -->
+
+We are going to use [logr](https://github.com/go-logr/logr) and
+[zap](https://github.com/uber-go/zap) (with
+[zapr](https://github.com/go-logr/zapr) as the shim between them) to construct
+the logging facilities.
+
+[logr](https://github.com/go-logr/logr) is a library that provides a common
```

Comment:
Because `controller-runtime` used it 😢 I think.

Reply:
Yes, `controller-runtime` introduces this `logr` to us. But the basic reason is `logr` is the "logging facade", it does not contain any implementation for logging, but only defines the interface for "Logger". 

The most benefits are:
- one day we would provide chaos mesh's component as the library, it would not depend on any **logging implementation/logging framework/logging library**;
- It means we could switch between many logging frameworks easily without any code changes;

Choosing "logging facade" would not bring many advantages now, the more consider things is the extensibility and flexible things. 🤔

About why `logr`, but not other libraries like https://github.com/go-eden/slf4go, because `controller-runtime` is using `logr`.  🤣